macaroonbakery/checker.py

```diff
@@ -36,4 +49,388 @@ def __hash__(self):
+        if caveats is None or len(caveats) == 0:
+            all_errors = []
+            all_errors.extend(self._init_errors)
+            all_errors.extend(errors)
+            err = ''
+            if len(all_errors) > 0:
+                err = ' '.join(all_errors)
+            raise PermissionDenied(err)
+        raise DischargeRequiredError(
+            msg='some operations have extra caveats', ops=ops, cavs=caveats)
+
+    def allow_capability(self, ctx, ops):
+        '''Checks that the user is allowed to perform all the
+        given operations. If not, a discharge error will be raised.
+        If allow_capability succeeds, it returns a list of first party caveat
+        conditions that must be applied to any macaroon granting capability
+        to execute the operations. Those caveat conditions will not
+        include any declarations contained in login macaroons - the
+        caller must be careful not to mint a macaroon associated
+        with the LOGIN_OP operation unless they add the expected
+        declaration caveat too - in general, clients should not create
+        capabilities that grant LOGIN_OP rights.
+
+        The operations must include at least one non-LOGIN_OP operation.
+        '''
+        nops = 0
+        for op in ops:
+            if op != LOGIN_OP:
+                nops += 1
+        if nops == 0:
+            raise ValueError('no non-login operations required in capability')
+
+        _, used = self._allow_any(ctx, ops)
+        squasher = _CaveatSquasher()
+        for i, is_used in enumerate(used):
+            if not is_used:
+                continue
+            for cond in self._conditions[i]:
+                squasher.add(cond)
+        return squasher.final()
+
+    def _check_conditions(self, ctx, op, conds):
+        declared = infer_declared_from_conditions(
+            conds,
+            self.parent.namespace())
+        ctx = context_with_operations(ctx, [op.action])
+        ctx = context_with_declared(ctx, declared)
+        for cond in conds:
+            err = self.parent._first_party_caveat_checker.\
+                check_first_party_caveat(ctx, cond)
+            if err is not None:
+                return None, err
+        return declared, None
+
+
+class AuthInfo(namedtuple('AuthInfo', 'identity macaroons')):
+    '''AuthInfo information about an authorization decision.
+    '''
+    __slots__ = ()
+
+    def __new__(cls, identity, macaroons):
+        '''
+        :param identity: holds information on the authenticated user as
+        returned identity_client. It may be None after a successful
+        authorization if LOGIN_OP access was not required.
+        :param macaroons: holds all the macaroons that were used for the
+        authorization. Macaroons that were invalid or unnecessary are
+        not included.
+        '''
+        return super(AuthInfo, cls).__new__(cls, identity, macaroons)
+
+
+class _CaveatSquasher(object):
```

Comment:
Please include a doc comment on this type.

Reply:
done.